proposals/NNNN-module-aliasing-for-disambiguation.md

```diff
@@ -0,0 +1,226 @@
+|— Module Utils (from package ‘swift-draw’) // want to add a new dependency
+```
+
+As an example, Utils from each package has the following code:
+```
+[Module Utils] // swift-game
+
+public struct Level { ... }
+public var currentLevel: Utils.Level { ... }
+```
+```
+[Module Utils] // swift-draw
+
+public protocol Drawable { ... }
+public class Canvas: Utils.Drawable { ... }
+```
+
+In the following steps, we will use a compiler invocation command to demonstrate how the conflicting modules can be aliased. 
+
+1. First, take the Utils module from swift-game, and rename it to be unique; let’s call it GameUtils. We will need to compile the module (a) by giving a new name GameUtils while (b) treating any references to Utils in its source files as GameUtils.  
+    a. The first part (renaming) can be achieved via passing the new name (GameUtils) to `-module-name` and an output path flag, `-o`,  `-emit-module-path`, or `-emit-module-interface-path`, which are all existing flags. The resulting binary then, for example, will be `/path/to/GameUtils.swiftmodule` (instead of `Utils.swiftmodule`).
+    b. The second part (treating references to Utils in source files as GameUtils) will be addressed by introducing a new compiler flag `-module-alias [name]=[binary_name]`, where the name is the module name that appears in source files (Utils), while the binary_name is the name of the .swiftmodule (GameUtils), the canonical name used for file contents including symbol mangling. Combined with step (a), the compiler invocation command will be `swiftc -module-name GameUtils -emit-module-path /path/to/GameUtils.swiftmodule -module-alias Utils=GameUtils ...`.    
+2. Similarly, we can rename Utils from swift-draw as DrawUtils, following the step 1.  
+3. Finally, we build App (without the aliasing flag) and directly import GameUtils and/or DrawUtils. 
+
+```
+[App]
+
+import GameUtils
+import DrawUtils
+```
+In practice, a dependency graph will be more complex than the scenario described above. As an example, we have the following modified scenario. App imports module Game, which imports module Utils from the same package. App also imports another module called Utils from a different package.  
+
+```
+App 
+|— Module Game (from package ‘swift-game’)
+|— Module Utils (from package ‘swift-game’)
+|— Module Utils (from package ‘swift-draw’)
+```
+
+```
+[Module Game] // swift-game
+
+import Utils  // swift-game
+public func start(level: Utils.Level) { ... }
+```
+
+As with the first scenario, the Utils modules are conflicting, so we need to perform the steps 1-2 above. Then we need to build module Game by applying `-module-alias Utils=GameUtils`, so that the references to Utils in source files of module Game are compiled as GameUtils without making any source changes. The compiler invocation command to build Game then is `swiftc -module-name Game -module-alias Utils=GameUtils ...`. App can then import module Game and any of the renamed Utils as needed. 
```

Comment:
Is it possible in this scenario to not rename the `Utils` from `swift-draw`?  If this conflict was introduced by updating `swift-game`, I imagine the developers of `App` would prefer to leave the `Utils` from `swift-draw` alone, since otherwise they will need to make a bunch of source changes.

Reply:
Yes it's possible; if a conflict is resolved by renaming one module, there's no need to alias the other module.

Reply:
Okay, makes sense.